Pure Underlay using KubeOVN Provider Networks (#1069)
harvester/docs: docs/networking/kubeovn-pureunderlay.md

:::

A pure underlay network allows virtual machines to connect directly to a physical Layer 2 network without overlay networking and the encapsulation overhead of technologies such as VXLAN or Geneve. Each virtual machine becomes a first-class member of the external network, communicating natively using standard Ethernet frames.

In Harvester, this direct connectivity is achieved by attaching virtual machines to a provider network, which maps a physical NIC or bonded interface on each host to the cluster. Traffic leaves the virtual machine through the host's physical interface and reaches the external network immediately, bypassing overlay tunnels completely.

![](/img/pureunderlay.png)

## Kube-OVN Underlay Mode

While traditional VLAN networking already enables virtual machines to communicate directly with the physical infrastructure, Kube-OVN's _Underlay_ mode extends this capability by embedding native Layer 2 connectivity directly into the OVN networking control plane. This deep integration allows workloads to leverage your existing physical underlay infrastructure while coexisting seamlessly with standard overlay networks inside the same Kubernetes cluster.

> Furthermore, integrating Kube-OVN’s underlay capabilities into Harvester ensures that virtual machines connected to the physical network do not lose their cloud-native security features. Workloads can still benefit from advanced micro-segmentation capabilities, including subnet ACLs and granular network policies, allowing you to maintain strict traffic control across both virtual and physical network boundaries.Use [Network Isolation](https://docs.harvesterhci.io/v1.9/networking/kubeovn-vm-isolation) to achieve microsegmentation of VMs using underlay Network.

Reviewer: ```suggestion
Furthermore, integrating Kube-OVN’s underlay capabilities into Harvester ensures that virtual machines connected to the physical network do not lose their cloud-native security features. Workloads can still benefit from advanced micro-segmentation capabilities, including [subnet ACLs and granular network policies](./kubeovn-vm-isolation.md), allowing you to maintain strict traffic control across both virtual and physical network boundaries.
```

> ### Underlay Installation

Reviewer: ```suggestion
## Underlay Installation
```

#### Create a Provider Network

1. Go to > **Underlay Networks**.

1. Select **Create**.

1. Configure the following settings:

    - Name

>     - Description (optional)

Reviewer: ```suggestion
### Create a Provider Network

1. On the Harvester UI, go to **Underlay Networks > Provider Networks**.

1. Click **Create**.

    ![](/img/pn.png)

1. Specify a unique name for the network.
```

1. On the **Interfaces** tab, configure the following settings:

    - Select **Default Interface**. This is any physical or bond interface available on all the Harvester host.The provider network using this spans all nodes in the cluster.
    - Select **Custom Interfaces**. (Optional). If the interface names are not same on all the nodes in the cluster, user provides specific interface name to nodename mapping.
      - Interface Name
      - Nodes
    - Select **Excluded Nodes**. (Optional). These nodes are excluded from provider network configuration.

>       - Nodes

Reviewer: ```suggestion
1. On the **Interfaces** tab, configure the following settings:

    - **Default Interface**: Select a physical or bonded interface that is present on every host in your cluster. This interface acts as the main uplink connecting all nodes to the provider network.

    - **Custom Interfaces**: (Optional) Configure a custom interface if your hosts do not have identical network interface names. You can manually map specific interface names to their corresponding nodes to ensure consistent network connectivity.

1. (Optional) On the **Excluded Nodes** tab, select nodes that should not participate in this provider network (for example, nodes that are isolated for dedicated tasks).
```

>     ![](/img/pn.png)

Reviewer: ```suggestion
1. Click **Create**.
```

Reviewer: I moved the image to the first step and added this missing final step.

> Refer [Provider Network Configuration](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-providernetwork)

Reviewer: ```suggestion
For more information, see [Create Provider Network](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-providernetwork) in the Kube-OVN documentation.
```

#### Create a VLAN Network

1. Go to > **VLANs**.

1. Select **Create**.

1. Configure the following settings:

    - Name

>     - Description (optional)

Reviewer: ```suggestion
### Create a VLAN Network

1. On the Harvester UI, go to **Underlay Networks > VLANs**.

    ![](/img/vlan2017.png)

1. Click **Create**.

1. Specify a unique name for the network.
```

1. On the **Basics** tab, configure the following settings:

    - Select **VLAN ID**.VLAN ID/Tag, Kube-OVN will add this Vlan tag to traffic, if set 0, no tag is added. the vlan tag applies to a localnet port.

>     - Select **Provider Network**. The name of ProviderNetwork. Multiple VLAN can use a same ProviderNetwork.

Reviewer: ```suggestion
1. On the **Basic** tab, configure the following settings:

    - **VLAN ID**: Specify the VLAN tag number for this network. Traffic passing through this network will automatically be tagged with this ID. If you do not want traffic to be tagged, set the value to `0`.

    - **Provider Network**: Select the underlying provider network. You can map multiple VLANs to the same provider network to segregate traffic over the same physical infrastructure.
```

>     ![](/img/vlan2017.png)

Reviewer: ```suggestion
1. Click **Create**.
```

> Refer [VLAN Network Configuration](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-vlan)

Reviewer: ```suggestion
For more information, see [Create VLAN](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-vlan) in the Kube-OVN documentation.
```

#### Create an Overlay Network

> Follow [Create an Overlay Network](https://docs.harvesterhci.io/v1.9/networking/harvester-network#create-an-overlay-network) to create an Overlay Network.

Reviewer: ```suggestion
### Create an Overlay Network

Follow the instructions in [Create an Overlay Network](./harvester-network.md#create-an-overlay-network).
```

1. Select `kube-system` namespace

1. On the **Basics** tab, configure the following additional settings:

    - Select **OverlayNetwork**.
    - Select **Network Interface Card** (`The NIC selected here must match the interface provided in Provider Network`)

>      ![](/img/vswitchexternal.png)

Reviewer: ```suggestion
![](/img/vswitchexternal.png)

:::info important

When setting up a pure underlay, the overlay network must use the following configuration:

- **Namespace**: Select `kube-system`.
- **Network Interface Card**: Select the identical physical or bonded network interfaces that you assigned when configuring the provider network.

:::
```

#### Create a Subnet in custom or default VPC

Follow [Create a Subnet](https://docs.harvesterhci.io/v1.8/networking/kubeovn-vpc/#subnet-settings)

1. On the **Basics** tab, configure the following additional settings:

    - **VLAN**. Name of the vlan resource configured for the Provider Network.

>      ![](/img/subnetexternal.png)

Reviewer: ```suggestion
### Create a Subnet in a VPC

For descriptions of settings that you must configure, see [Subnet Settings](./kubeovn-vpc.md#subnet-settings).

![](/img/subnetexternal.png)

:::info important

When setting up a pure underlay, you must specify the name of the VLAN network that is mapped to the provider network.

:::
```

#### Create VMs attached to underlay network using `vswitchexternal`

> Refer [Create a VM](https://docs.harvesterhci.io/v1.9/vm/index#how-to-create-a-vm)

Reviewer: Is this section necessary? The title of the parent section is "Underlay Installation". VM creation seems external to the installation process.

Author: It is better if users are aware that they could create VMs over the underlay.
The users after successful installation of underlay can spin up VMs and attach the VMs to the underlay network's VM Network/provider and VMs can communicate to other VMs or to external using the underlay network.

Author: I can add section or rephrase it based on your suggestion.

Reviewer: See the text I suggested.

![](/img/pureunderlay.png)

## Kube-OVN Underlay Mode

While traditional VLAN networking already enables virtual machines to communicate directly with the physical infrastructure, Kube-OVN's _Underlay_ mode extends this capability by embedding native Layer 2 connectivity directly into the OVN networking control plane. This deep integration allows workloads to leverage your existing physical underlay infrastructure while coexisting seamlessly with standard overlay networks inside the same Kubernetes cluster.

Furthermore, integrating Kube-OVN’s underlay capabilities into Harvester ensures that virtual machines connected to the physical network do not lose their cloud-native security features. Workloads can still benefit from advanced micro-segmentation capabilities, including [subnet ACLs and granular network policies](./kubeovn-vm-isolation.md), allowing you to maintain strict traffic control across both virtual and physical network boundaries.

> ## Underlay Installation

Reviewer: ```suggestion
## Underlay Configuration
```

1. Click **Create**.

For more information, see [Create VLAN](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-vlan) in the Kube-OVN documentation.

### Create an Overlay Network

Follow the instructions in [Create an Overlay Network](./harvester-network.md#create-an-overlay-network).

![](/img/vswitchexternal.png)

:::info important

When setting up a pure underlay, the overlay network must use the following configuration:

- **Namespace**: Select `kube-system`.
- **Network Interface Card**: Select the identical physical or bonded network interfaces that you assigned when configuring the provider network.

:::

### Create a Subnet in a VPC

For descriptions of settings that you must configure, see [Subnet Settings](./kubeovn-vpc.md#subnet-settings).

![](/img/subnetexternal.png)

:::info important

When setting up a pure underlay, you must specify the name of the VLAN network that is mapped to the provider network.

:::

#### Create VMs attached to underlay network using `vswitchexternal`

> Refer [Create a VM](https://docs.harvesterhci.io/v1.9/vm/index#how-to-create-a-vm)

Reviewer: ```suggestion
## Connecting Virtual Machines to the Underlay

You can create virtual machines and attach them directly to a VLAN Network mapped to the configured provider network. These virtual machines can communicate with other virtual machines on the same subnet and route traffic directly to external networks via the physical infrastructure.

For more information, see [Create a Virtual Machine](../vm/create-vm.md#how-to-create-a-vm).
```

---
sidebar_position: 12
sidebar_label: Kube-OVN Pure Underlay

> title: "Pure Underlay Networking Using Kube-OVN Provider Networks"

Reviewer: ```suggestion
sidebar_label: Pure Underlay Networking
title: "Pure Underlay Networking"
```

keywords:
- Harvester
- networking
- Kube-OVN
- overlay VMs
- underlay networking
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.9/networking/kubeovn-pureunderlay"/>
</head>

:::note

All features that use Kube-OVN are considered experimental. For more information about experimental features, see [Feature Labels](../getting-started/document-conventions.md#feature-labels).

:::

> Pure Underlay Network allows virtual machines (VMs) to connect directly to the physical Layer 2 network, without encapsulation (such as VXLAN/Geneve) or overlay networking. Each VM becomes a first-class member of the external network and communicates using native Ethernet frames.

Reviewer: ```suggestion
A pure underlay network allows virtual machines to connect directly to a physical Layer 2 network without overlay networking and the encapsulation overhead of technologies such as VXLAN or Geneve. Each virtual machine becomes a first-class member of the external network, communicating natively using standard Ethernet frames.
```

> In Harvester using KubeOVN, this is typically achieved by attaching VMs to a Provider Network, which maps a physical NIC (or bonded interface) on each host into the cluster. Traffic leaves the VM through the host's physical interface and reaches the external network without passing through an overlay tunnel.

Reviewer: ```suggestion
In Harvester, this direct connectivity is achieved by attaching virtual machines to a provider network, which maps a physical NIC or bonded interface on each host to the cluster. Traffic leaves the virtual machine through the host's physical interface and reaches the external network immediately, bypassing overlay tunnels completely.
```

> ##  Why Pure underlay ?

Reviewer: ```suggestion
## Kube-OVN Underlay Mode
```

> While traditional VLAN networking already enables VMs to communicate directly with the physical network, Kube-OVN's Pure Underlay extends this capability by integrating native Layer 2 connectivity into the Kube-OVN networking model. This allows workloads to leverage existing underlay infrastructure while coexisting seamlessly with overlay networks in the same Kubernetes cluster.

Reviewer: ```suggestion
While traditional VLAN networking already enables virtual machines to communicate directly with the physical infrastructure, Kube-OVN's _Underlay_ mode extends this capability by embedding native Layer 2 connectivity directly into the OVN networking control plane. This deep integration allows workloads to leverage your existing physical underlay infrastructure while coexisting seamlessly with standard overlay networks inside the same Kubernetes cluster.
```

> Also by integrating underlay networking of Kube-OVN into SUSE Virtualization, VMs connected to the physical network can still benefit from Kube-OVN's micro-segmentation capabilities, including Subnet ACLs and Network Policies. This enables fine-grained traffic control between workloads.

Reviewer: ```suggestion
Furthermore, integrating Kube-OVN’s underlay capabilities into Harvester ensures that virtual machines connected to the physical network do not lose their cloud-native security features. Workloads can still benefit from advanced micro-segmentation capabilities, including subnet ACLs and granular network policies, allowing you to maintain strict traffic control across both virtual and physical network boundaries.
```

> ### Underlay Installation

Reviewer: The procedures don't provide actual configuration guidance. Instead of telling users to check external documentation, can we add to our doc the key pieces of information that will help them complete the procedure?

Also, should users complete the following procedures in a specific order?

Author: Ok, so for provider network and vlan network I will add details of configuration in this doc.Do you also want me to remove the external links from kubeovn ? 

And for create VM section, is the link to our documentation enough ?

The steps go in same order as its explained in the doc.

#### Create a Provider Network

1. Go to > **Underlay Networks**.

1. Select **Create**.

1. Configure the following settings:

    - Name
    - Description (optional)

1. On the **Interfaces** tab, configure the following settings:

    - Select **Default Interface**.
    - Select **Custom Interfaces**. (Optional)
      - Interface Name
      - Nodes
    - Select **Excluded Nodes**. (Optional)

>       - Nodes

Reviewer: I'll review the wording in this section after we clarify some details.

- How should users choose an interface from the ones listed in the **Default Interface** drop-down list?
- Why should users configure a custom interface? Provide typical reasons/use cases.
- When should users exclude nodes? Provide typical reasons/use cases.

Author: I will add those details to the document for you to review, thanks

    ![](/img/pn.png)

Refer [Provider Network Configuration](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-providernetwork)

#### Create a VLAN Network

1. Go to > **VLANs**.

1. Select **Create**.

1. Configure the following settings:

    - Name
    - Description (optional)

1. On the **Basics** tab, configure the following settings:

    - Select **VLAN ID**.
    - Select **Provider Network**.

    ![](/img/vlan2017.png)

Refer [VLAN Network Configuration](https://kubeovn.github.io/docs/v1.16.x/en/start/underlay/#create-vlan)

> #### Create an Overlay Network

Reviewer: We can just link to https://docs.harvesterhci.io/v1.8/networking/harvester-network#create-an-overlay-network and highlight the settings that are specific to the user's goal in this context. This is assuming that the linked part is up-to-date.

Author: For creating an overlay network, there are two new fields (kube-system namespace and NIC) which is not available in the older doc.

1. Go to **Networks** > **Overlay Networks**.

1. Select **Create**.

1. Configure the following settings:

    - Namespace (`kube-system`)
    - Name
    - Description (optional)

1. On the **Basics** tab, configure the following settings:

    - Select **OverlayNetwork**.
    - Select **Network Interface Card** (`The NIC selected here must match the interface provided in Provider Network`)

     ![](/img/vswitchexternal.png)

#### Create a Subnet in custom or default VPC

1. Go to **Overlay Networks** > **Virtual Private Cloud**.

> 1. Select **Create Subnet** under `ovn-cluster`

Reviewer: The section title says custom or default VPC, but step 2 tells the user to create the subnet in `ovn-cluster`. Is that required or just an example?

Author: that is just an example

1. Configure the following settings:

    - Name
    - Description (optional)

1. On the **Basics** tab, configure the following settings:

    - **CIDR Block**.
    - **Provider Network**.
    - **Gateway IP**
    - **Protocol**
    - **VPC**

>     - **VLAN**

Reviewer: What about the other three settings on the **Basic** tab? The doc should be clear about whether users can safely keep the defaults or select different values.

https://docs.harvesterhci.io/v1.8/networking/kubeovn-vpc/#subnet-settings - We can link to this.

Author: Other settings (like private subnet,allowed subnets) are optional and there is a new field vlan added in v1.9 which is not available in older v1.18 subnet creation document

     ![](/img/subnetexternal.png)

#### Create VMs attached to underlay network using `vswitchexternal`
Refer [Create a VM](https://docs.harvesterhci.io/v1.9/vm/index#how-to-create-a-vm)

### Underlay Networking Explained

> ![](/img/pureunderlay.png)

Reviewer: The diagram belongs to the intro section and doesn't need its own heading.

Author: ok, I will move the diagram to the intro section

### Micro segmentation

> Use [Network Isolation](https://docs.harvesterhci.io/v1.9/networking/kubeovn-vm-isolation) to achieve microsegmentation of VMs using underlay Network.

Reviewer: Since micro-segmentation was already mentioned in the intro section, we can add the link there and remove this section altogether.

Author: ok, I will change this
